### ISSUE TITLE
Should warn when trying to cast between a fat pointer and a non-fat pointer

### ISSUE THREAD

**Issue by upsuper** [rust-lang/rust#64806]
Please see the following piece of code (I came up with when discussing in our Rust group):
```rust
use std::boxed::Box;
use libc::c_void;

fn main() {
    show_ptr(get_ptr());
}

fn get_ptr() -> *mut c_void {
    let x = vec!["Hello".to_string(), "World".to_string()].into_boxed_slice();
    Box::into_raw(x) as *mut c_void
}

fn show_ptr(ptr: *mut c_void) {
     let x = unsafe { Box::from_raw(ptr as *mut Box<[String]>) };
     println!("{:?}", x);
}
```

It builds fine but confusingly segfaults. The reason is that boxed slice and its pointer (`*mut [String]`) are fat pointers, and casting it to `*mut c_void` is losing information.

I believe there should at least be a compiler warning for this kind of conversion, since it is very likely to be broken.

**Comment by upsuper** [comment]
In this specific case, I think the compiler may encourage people to go through `Vec` in order to get both pointer and length.

Rust can probably also add support to cast this kind of fat pointer to `(*mut T, usize)`.

**Comment by upsuper** [comment]
Hmmm... `*mut Box<[String]>` is not a fat pointer, so the conversion back doesn't have problem. This issue is probably not valid.
